Bottom: 983a145a01c949fba7379a35e1841c79f3703fd9
Top:    c2e27822411312b615c5f7737fdfc2b41248a84c
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-02-15 13:45:16 +0800

virtio-net: xdp xmit more

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 42cbf9f..9876adb 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -86,6 +86,8 @@ struct send_queue {
 
 	/* Name of the send queue: output.$index */
 	char name[40];
+
+	bool xdp_xmit_more;
 };
 
 /* Internal representation of a receive virtqueue */
@@ -378,7 +380,11 @@ static bool virtnet_xdp_xmit(struct virtnet_info *vi,
 		return false;
 	}
 
-	virtqueue_kick(sq->vq);
+	if (sq->vq->num_free < 2)
+		virtqueue_kick(sq->vq);
+	else
+		sq->xdp_xmit_more = true;
+
 	return true;
 }
 
@@ -1005,7 +1011,9 @@ static int virtnet_poll(struct napi_struct *napi, int budget)
 {
 	struct receive_queue *rq =
 		container_of(napi, struct receive_queue, napi);
-	unsigned int r, received;
+	struct virtnet_info *vi = rq->vq->vdev->priv;
+	struct send_queue *sq;
+	unsigned int r, received, qp;
 
 	received = virtnet_receive(rq, budget);
 
@@ -1021,6 +1029,16 @@ static int virtnet_poll(struct napi_struct *napi, int budget)
 		}
 	}
 
+	if (rq->xdp_prog) {
+		qp = vi->curr_queue_pairs - vi->xdp_queue_pairs +
+		     smp_processor_id();
+		sq = &vi->sq[qp];
+		if (sq->xdp_xmit_more) {
+			sq->xdp_xmit_more = false;
+			virtqueue_kick(sq->vq);
+		}
+	}
+
 	return received;
 }
